JsonSubTypes/JsonSubtypes.cs

```diff
@@ -241,11 +241,16 @@ private Type GetType(JObject jObject, Type parentType, JsonSerializer serializer
             Type targetType = parentType;
             JsonSubtypes lastTypeResolver = null;
             JsonSubtypes currentTypeResolver = this;
+            HashSet<Type> visitedTypes = new HashSet<Type> {targetType};
 
             var jsonConverterCollection = serializer.Converters.OfType<JsonSubtypesConverter>().ToList();
             while (currentTypeResolver != null && currentTypeResolver != lastTypeResolver)
```

Comment:
Why not simply add the condition ` && jsonConverterCollection.Count > 0` ?

Reply:
I didn't have a good enough understanding of the code and loop invariant here to recognize that as a potential solution. I just saw a problem with repeatedly resolving the same target type, and implemented the most straightforward solution I could see to that.

Just adding `&& jsonConverterCollection.Count > 0` to the loop condition seems to cause the `ConcurrentThreadTest` and `Test` tests in `BaseIsAnInterfaceTests`, the `Demo` test in `DemoCustomSubclassMappingTests`, and the `Demo` and `DemoDifferentCase` tests in `DemoKnownSubTypeWithProperty` to fail. However, then changing it to a do while loop fixes that.

Would you prefer that as a solution? I can commit it to my fork or you can commit it there yourself (I have allowed edits from maintainers).

Reply:
no, let's use the HashSet to keep track of visited types

Reply:
I think this solution is a better one:
![image](https://user-images.githubusercontent.com/5827164/63223653-88381c80-c1b9-11e9-962c-1d782c8eb2af.png)

It ensures that there is a progression in the loop without using a set

The proposed check I made on `jsonConverterCollection` is not correct because there might left some `JsonSubtypesConverter` in the list that does not apply to this type hierarchy

Reply:
You can still mess that up with a circular reference in three or more classes:

``` cs
[JsonConverter(typeof(JsonSubtypes), "Discriminator")]
[JsonSubtypes.KnownSubType(typeof(B), "D")]
public abstract class A
{
}

[JsonConverter(typeof(JsonSubtypes), "Discriminator")]
[JsonSubtypes.KnownSubType(typeof(C), "D")]
public abstract class B
{
}

[JsonConverter(typeof(JsonSubtypes), "Discriminator")]
[JsonSubtypes.KnownSubType(typeof(A), "D")]
public abstract class C
{
}

JsonConvert.DeserializeObject<A>("{\"Discriminator\":\"D\"}");
```

You may say that you're pointing the gun at your own foot at that point, but could it happen inadvertently?

Reply:
Let's keep the HashSet

Reply:
Would you like the above A-B-C circle as a test as well?

Reply:
I have pushed the A-B-C test as well.